earth_enterprise/src/fusion/autoingest/AssetFactory.h

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2019 The Open GEE Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef ASSET_FACTORY_H
+#define ASSET_FACTORY_H
+
+#include <common/SharedString.h>
+#include <common/khMetaData.h>
+
+// The goal of this namspace is to have a single place in the code where all
+// asset creation is handled.
+namespace AssetFactory
+{
+  template<class AssetType>
+  AssetType Find(const std::string & ref, const AssetDefs::Type & type)
+  {
+    assert(type != AssetDefs::Invalid);
+    const std::string subtype = AssetType::Impl::EXPECTED_SUBTYPE;
+    try {
+      AssetType asset(ref);
+      if (asset &&
+          (asset->type == type) &&
+          (asset->subtype == subtype)) {
+          return asset;
+      }
+    } catch (...) {
+        // do nothing - don't even generate any warnings
+    }
+    return AssetType();
+  }
+
+  template<class AssetType>
+  AssetType Find(const std::string & ref) {
+    return Find<AssetType>(ref, AssetType::Impl::EXPECTED_TYPE);
+  }
+
+  template<class VersionType>
+  void ValidateRefForInput(const std::string & ref, const AssetDefs::Type & type)
+  {
+    assert(type != AssetDefs::Invalid);
+    using AssetType = typename VersionType::Impl::AssetType;
+    const std::string subtype = VersionType::Impl::EXPECTED_SUBTYPE;
+    if (AssetVersionRef(ref).Version() == "current") {
+      AssetType asset = Find<AssetType>(ref, type);
+      if (!asset) {
+        throw std::invalid_argument(
+            "No such " + ToString(type) + " " + subtype + " asset: " + ref);
+      }
+    } else {
+      VersionType version = Find<VersionType>(ref, type);
+      if (!version) {
+        throw std::invalid_argument(
+            "No such " + ToString(type) + " " + subtype + " asset version: " +
+            ref);
+      }
+    }
+  }  
+  
+  template<class VersionType>
+  void ValidateRefForInput(const std::string & ref) {
+    return ValidateRefForInput<VersionType>(ref, VersionType::Impl::EXPECTED_TYPE);
+  }
+ 
+  template<class MutableDerivedAssetHandleType, class ConfigType>
```

Comment:
There's a potential problem here where a user could pass a `MutableDerivedAssetHandleType` that doesn't match the `ConfigType`. You can get around that by doing something similar to what you did for `AssetStorageType`. If each class has a `using` statement for its config type, then you only have to pass the handle type.

Reply:
Also, there's an opportunity to combine these 3 functions. You can have one function that takes all the possible parameters and then have the other two functions call that function, passing in default parameters where appropriate. I'm not sure that saves us much since the function body is effectively one line, though.

Reply:
> Also, there's an opportunity to combine these 3 functions.
I initially went that route, but I ran into a seg fault that seemed to have to do with having `const std::vector<SharedString>& inputs_ = std::vector<SharedString>()` and then passing `inputs_` into `AssetStorageType::MakeStorage`.
Maybe I'm remembering that incorrectly. I can try it again to see if that works.

Reply:
It's not enough duplication that I'm very concerned about it. I'm OK either way.

Reply:
I started looking again at trying to consolidate them. It'd take a little reworking of the auto-generated function arguments since the default parameters would need to be at the end. And some of those auto-gen arguments include commas, some don't. And they're used other places besides these functions. It looks likely to be error-prone and more work than I'd hope.
At this point, I think it's best to leave it as is.